pkg/drivers/response.go

```diff
@@ -0,0 +1,103 @@
+package drivers
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/MontFerret/ferret/pkg/runtime/core"
+	"github.com/MontFerret/ferret/pkg/runtime/values"
+	"github.com/MontFerret/ferret/pkg/runtime/values/types"
+)
+
+// HTTPResponse HTTP response object.
+type HTTPResponse struct {
+	StatusCode int
+	Status     string
+	Headers    HTTPHeaders
+}
+
+func (resp *HTTPResponse) Type() core.Type {
+	return HTTPResponseType
+}
+
+func (resp *HTTPResponse) String() string {
+	return resp.Status
+}
+
+func (resp *HTTPResponse) Compare(other core.Value) int64 {
+	if other.Type() != HTTPResponseType {
+		return Compare(HTTPResponseType, other.Type())
+	}
+
+	otherObj := values.Parse(other)
+	respObj := values.Parse(resp)
+
+	// HTTPHeaders has it's own Compare
+	respObj.(*values.Object).Set(
+		values.NewString("Headers"),
+		resp.Headers,
+	)
+	otherObj.(*values.Object).Set(
+		values.NewString("Headers"),
+		other.(*HTTPResponse).Headers,
+	)
+
+	return respObj.Compare(otherObj)
+}
+
+func (resp *HTTPResponse) Unwrap() interface{} {
+	return resp
+}
+
+func (resp *HTTPResponse) Copy() core.Value {
+	return *(&resp)
+}
+
+func (resp *HTTPResponse) Hash() uint64 {
+	return values.Parse(resp).Hash()
+}
+
+// responseMarshal is a structure that repeats HTTPResponse. It allows
+// easily Marshal the HTTPResponse object.
+type responseMarshal struct {
```

Comment:
Why did you decide to go with this approach? Did you consider other options?

For example, using map[string]interface{} for marshalling in ``HTTPResponse`` struct?

```go
func (resp *HTTPResponse) MarshalJSON() ([]byte, error) {
    j := map[string]interface{}{
        "status_code": resp.StatusCode,
        "status": resp.Status,
        "headers": resp.Headers
    }
    
    return json.Marshal(j)
}
```

Or, maybe embedding would work?

```go
type HTTPResponseBase struct {
	StatusCode int         `json:"status_code"`
	Status     string      `json:"status"`
	Headers    HTTPHeaders `json:"headers"`
}

type HTTPResponse struct {
	HTTPResponseBase
}

func (resp *HTTPResponse) MarshalJSON() ([]byte, error) {
	if resp == nil {
		return json.Marshal(values.None)
	}

	return json.Marshal(resp.HTTPResponseBase)
}

```

Reply:
I decided to use this approach because in my opinion this is best way to keep HTTPResponse and it's marshal form consistent.

1. Why not using map. Because if another field appears in the HTTPResponse, it’s easy to forget to add it to the map.

2. Why not embedding. Embedded struct this is the same with current approach, but have a single difference.
```go
// Init using different struct
r := drivers.HTTPResponse{
    StatusCode: resp.StatusCode,
    Status: resp.Status,
    Headers: drivers.HTTPHeaders(resp.Header),
}

// Init using embedded struct - got one more nested level
r := drivers.HTTPResponse{
    HTTPResponseBase: drivers.HTTPResponseBase{
        StatusCode: resp.StatusCode,
        Status: resp.Status,
        Headers: drivers.HTTPHeaders(resp.Header),
    }
}
```

Reply:
Code will not compile if `HTTPResponse` and `responseMarshal` would have different fields. 
```go
// there will be a compile error
return json.Marshal(responseMarshal(*resp))
```

Reply:
> I decided to use this approach because in my opinion this is best way to keep HTTPResponse and it's marshal form consistent.
> 
> 1. Why not using map. Because if another field appears in the HTTPResponse, it’s easy to forget to add it to the map.
> 2. Why not embedding. Embedded struct this is the same with current approach, but have a single difference.
> 
> ```go
> // Init using different struct
> r := drivers.HTTPResponse{
>     StatusCode: resp.StatusCode,
>     Status: resp.Status,
>     Headers: drivers.HTTPHeaders(resp.Header),
> }
> 
> // Init using embedded struct - got one more nested level
> r := drivers.HTTPResponse{
>     HTTPResponseBase: drivers.HTTPResponseBase{
>         StatusCode: resp.StatusCode,
>         Status: resp.Status,
>         Headers: drivers.HTTPHeaders(resp.Header),
>     }
> }
> ```

I do not see a problem in a nested struct - we have just 2 places where it will be used and it will unlikely change. But the benefit is that you don't have to have 2 structs with identical fields.

Reply:
`HTTPResponseBase` implies that somewhere there must be `HTTPResponse` or `HTTPResponseExtended` or something like that. 

This approach provides more clear and small package API. Keeping both `HTTPResponseBase` and `HTTPResponse` can be confusing.